### ISSUE TITLE
DB4S compile failure on OSX 10.11.6 with XCode 8

### ISSUE THREAD

**Issue by justinclift** [sqlitebrowser/sqlitebrowser#776]
Updated my system last night with "XCode 8", without thinking it though. 😦 

Now, DB4S compile fails when used with Qt5:

```
/Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/clang++ -c -pipe -stdlib=libc++ -std=c++11 -DENABLE_SQLCIPHER -DCHECKNEWVERSION -O2 -std=gnu++11 -isysroot /Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX10.12.sdk -mmacosx-version-min=10.8 -Wall -W -fPIC -DQT_NO_DEBUG -DQT_PRINTSUPPORT_LIB -DQT_WIDGETS_LIB -DQT_MACEXTRAS_LIB -DQT_GUI_LIB -DQT_NETWORK_LIB -DQT_CORE_LIB -I. -I/usr/local/opt/sqlcipher/include -isystem /usr/local/include -I../libs/antlr-2.7.7 -I../libs/qhexedit -I../libs/qcustomplot-source -I../libs/qscintilla/Qt4Qt5 -I../../sqlitebrowser -I../../../Qt/5.7/clang_64/lib/QtPrintSupport.framework/Headers -I../../../Qt/5.7/clang_64/lib/QtWidgets.framework/Headers -I../../../Qt/5.7/clang_64/lib/QtMacExtras.framework/Headers -I../../../Qt/5.7/clang_64/lib/QtGui.framework/Headers -I../../../Qt/5.7/clang_64/lib/QtNetwork.framework/Headers -I../../../Qt/5.7/clang_64/lib/QtCore.framework/Headers -Irelease -I/Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX10.12.sdk/System/Library/Frameworks/OpenGL.framework/Headers -I/Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX10.12.sdk/System/Library/Frameworks/AGL.framework/Headers -I.ui -I../../../Qt/5.7/clang_64/mkspecs/macx-clang -F/Users/jc/Qt/5.7/clang_64/lib -o release/PreferencesDialog.o PreferencesDialog.cpp
In file included from MainWindow.cpp:2:
In file included from .ui/ui_MainWindow.h:43:
../libs/qcustomplot-source/qcustomplot.h:448:26: error: implicit instantiation of undefined template 'QPointer<QCPLayerable>'
  QPointer<QCPLayerable> mParentLayerable;
                         ^
../../../Qt/5.7/clang_64/lib/QtCore.framework/Headers/qmetatype.h:1339:1: note: template is declared here
QT_FOR_EACH_AUTOMATIC_TEMPLATE_SMART_POINTER(QT_FORWARD_DECLARE_SHARED_POINTER_TYPES_ITER)
^
../../../Qt/5.7/clang_64/lib/QtCore.framework/Headers/qmetatype.h:215:7: note: expanded from macro 'QT_FOR_EACH_AUTOMATIC_TEMPLATE_SMART_POINTER'
    F(QPointer)
      ^
In file included from MainWindow.cpp:2:
In file included from .ui/ui_MainWindow.h:43:
../libs/qcustomplot-source/qcustomplot.h:1465:21: error: implicit instantiation of undefined template 'QPointer<QCPAxis>'
  QPointer<QCPAxis> mKeyAxis, mValueAxis;
                    ^
../../../Qt/5.7/clang_64/lib/QtCore.framework/Headers/qmetatype.h:1339:1: note: template is declared here
QT_FOR_EACH_AUTOMATIC_TEMPLATE_SMART_POINTER(QT_FORWARD_DECLARE_SHARED_POINTER_TYPES_ITER)
^
../../../Qt/5.7/clang_64/lib/QtCore.framework/Headers/qmetatype.h:215:7: note: expanded from macro 'QT_FOR_EACH_AUTOMATIC_TEMPLATE_SMART_POINTER'
    F(QPointer)
      ^
In file included from MainWindow.cpp:2:
In file included from .ui/ui_MainWindow.h:43:
../libs/qcustomplot-source/qcustomplot.h:1465:31: error: implicit instantiation of undefined template 'QPointer<QCPAxis>'
  QPointer<QCPAxis> mKeyAxis, mValueAxis;
                              ^
../../../Qt/5.7/clang_64/lib/QtCore.framework/Headers/qmetatype.h:1339:1: note: template is declared here
QT_FOR_EACH_AUTOMATIC_TEMPLATE_SMART_POINTER(QT_FORWARD_DECLARE_SHARED_POINTER_TYPES_ITER)
^
../../../Qt/5.7/clang_64/lib/QtCore.framework/Headers/qmetatype.h:215:7: note: expanded from macro 'QT_FOR_EACH_AUTOMATIC_TEMPLATE_SMART_POINTER'
    F(QPointer)
      ^
```

It's compiling and running perfectly fine against Qt4 though.

Anyone have suggestion on what it could be caused by?

**Comment by justinclift** [comment]
Removed Xcode 8, and re-installed Xcode 7.3.1.  DB4S compiles without issue.

Looks like an Xcode 8 specific problem.

**Comment by revolter** [comment]
[This answer][1] says that it will be fixed in Qt 5.6.1, but it offers no source.

  [1]: http://stackoverflow.com/a/39499500/865175

**Comment by justinclift** [comment]
Nah, that answer is for a different Xcode 8 problem.  It's for an incompatibility with Qt 5.6.0/5.7.0 and Xcode 8.

When starting Qt Creator 5.7.0 with Xcode 8 installed.  This error is displayed:

```
Project ERROR: Xcode not set up properly. You may need to confirm the license agreement
by running /usr/bin/xcodebuild.
```

The link you gave has a working fix.

The compile problem I'm reporting here isn't fixed by that. 😦

**Comment by joshua-meng** [comment]
Here is the solution:
http://stackoverflow.com/questions/33728905/qt-creator-project-error-xcode-not-set-up-properly-you-may-need-to-confirm-t

**Comment by justinclift** [comment]
@joshua-meng Hmmm, that didn't work for me.  I had already made that fix to get past the initial Xcode 8 warning.

Is it working for you?

**Comment by joshua-meng** [comment]
@justinclift 
I followed the steps then Qt Creator can create & compile GUI project on OSX 10.11.6 and Qt5.7.0.

>= Xcode 8

In Xcode 8, as Bruce said, this happens when Qt tries to find xcrun when it should be looking for xcodebuild.

Step 1:

Open the file:

Qt_install_folder/5.7/clang_64/mkspecs/features/mac/default_pre.prf
Step 2:

Replace:

isEmpty($$list($$system("/usr/bin/xcrun -find xcrun 2>/dev/null"))))
With:

isEmpty($$list($$system("/usr/bin/xcrun -find xcodebuild 2>/dev/null")))

**Comment by justinclift** [comment]
Thanks.  I'll look back into this again later, I must have been doing something wrong first time around. :smile:

**Comment by joshua-meng** [comment]
@justinclift That steps only makes Qt5.7.0 works with XCode8 on helloworld like applications, QScintilla still cannot compile. I got errors like:
```
/Users/joshua/programs/Qt5.7.0/5.7/clang_64/lib/QtCore.framework/Headers/qbasicatomic.h:61:4: error: 
      "Qt requires C++11 support"
#  error "Qt requires C++11 support"
   ^
/Users/joshua/programs/Qt5.7.0/5.7/clang_64/lib/QtCore.framework/Headers/qbasicatomic.h:90:13: error: 
      unknown type name 'QAtomicOps'
    typedef QAtomicOps<T> Ops;
            ^
/Users/joshua/programs/Qt5.7.0/5.7/clang_64/lib/QtCore.framework/Headers/qbasicatomic.h:90:23: error: 
      expected member name or ';' after declaration specifiers
    typedef QAtomicOps<T> Ops;

```

**Comment by justinclift** [comment]
Yeah.  Use Xcode 7.3.1 instead for now.  Not a great solution, but it works. :smile:

**Comment by ilovezfs** [comment]
@justinclift Mind giving this patch to qscintilla2 a try?
https://gist.github.com/ilovezfs/dff9d68264e513419e8830a804acb0b5

**Comment by justinclift** [comment]
Sure, will do.  Will get the Xcode 8 install happening again shortly to test with. :smile:

**Comment by revolter** [comment]
I'll test this tonight too.

**Comment by ilovezfs** [comment]
Thanks!!!

**Comment by revolter** [comment]
@ilovezfs, I guess that it worked for you with that patch, right?

**Comment by justinclift** [comment]
@revolter This is the patch from @ilovezfs, adjusted to work with our QScintilla: https://github.com/justinclift/sqlitebrowser/commit/4590ff6bf749ca389e66423740c0c199d1b1492d

**Comment by ilovezfs** [comment]
Yes, it got me all the way to the Octave GUI built under 10.12 with Xcode 8 and Qt 5.6.1 and 5.7.0 (with relevant Qt patches too).

**Comment by justinclift** [comment]
@ilovezfs Any idea if it's upstream, or been submitted there?  @revolter Asked on their mailing list in regards to XCode8 compile troubles, and no-one bothered to respond. 😦

**Comment by ilovezfs** [comment]
Nope, I've not poked qscintilla upstream yet with my patch. I'm easily confused when upstreams aren't on GitHub :smiling_imp:

The "revelation" was that all Qt5 reverse dependencies seemed fine except for qscintilla2 so it was actually C++ issue not a Qt5 issue at all.

**Comment by justinclift** [comment]
Heh Heh Heh 😈

**Comment by revolter** [comment]
So if it's a C++ issue and not related to Xcode 8, why does it build in Linux?

**Comment by justinclift** [comment]
XCode 8 probably has a different level of C++ support than XCode 7.x.

To investigate it more closely, we could check what version of clang is bundled with each, and compare C++ support reported by those, then check what your compiler on Linux says it supports.

**Comment by ilovezfs** [comment]
It's most assuredly an Xcode 8 issue (happens on 10.11 and 10.12 with Xcode 8), but it's not a Qt5 issue other than incidentally.

**Comment by ilovezfs** [comment]
@justinclift yes that sounds like a good idea. For instance, Xcode 8 supports more of the c++11 standard.

**Comment by justinclift** [comment]
Yep, it's compiling here perfectly now. :smile:

Tested with XCode 8 on both Qt4 and Qt5.

I'll merge the patch into both master and our v3.9.x branches shortly.

**Comment by justinclift** [comment]
@ilovezfs Thank you heaps. 😀

**Comment by revolter** [comment]
Awesome! Do you think it's a good idea to update `QScintilla` before applying the patch?

**Comment by ilovezfs** [comment]
It would be good to verify the patch doesn't break anything that was working on <= Xcode 7 x <= 10.11, so that we know whether it has to be constrained in any way, but ultimately that's qscintilla2's burden to bear. (e.g., does the patch break Yosemite + Xcode 6 or 7, etc.)

**Comment by justinclift** [comment]
Hmmm, good point.  I'll test with XCode 7.3.1 again shortly. :smile:

**Comment by justinclift** [comment]
@revolter Sure.  Probably make it a PR though instead of directly pushing it, so we can do a bit of testing on it first.

**Comment by revolter** [comment]
I tried updating it hoping it will solve the build problem and I remember it updated with no conflicts or anything.

**Comment by justinclift** [comment]
@ilovezfs Just tested it here with Xcode 7.3.1 as well.  Still compiling fine with both Qt4 and Qt5.

Looks like a good solution. :smile:

Btw, this is the commit applied to both our active branches: https://github.com/sqlitebrowser/sqlitebrowser/commit/969e263e5bf71e63df8ee908cb4a2014b9a55c9b

**Comment by justinclift** [comment]
@revolter Sounds good. :smile:

**Comment by revolter** [comment]
I can't believe this works, and I can finally build it again, and it was such a small fix. Thank you a lot, @ilovezfs!

**Comment by ilovezfs** [comment]
You're welcome :) Now we can upgrade Qt5 in Homebrew to support Sierra with a clear conscience hehe

**Comment by revolter** [comment]
@ilovezfs, btw, ilovebrew! I really do <3

**Comment by justinclift** [comment]
@ilovezfs Do you want to submit the patch to upstream QScintilla?  If you're feeling lazy, we can do it. :smile:

**Comment by ilovezfs** [comment]
@justinclift that would be awesome. There's little I like less than "reporting upstream"

**Comment by justinclift** [comment]
@revolter You up for it, or should I?

**Comment by ilovezfs** [comment]
In case this helps: http://stackoverflow.com/questions/15583604/implicit-instantiation-of-undefined-template

**Comment by revolter** [comment]
@justinclift I have no idea where the QScintilla source code is

**Comment by ilovezfs** [comment]
I think someone may keep it printed out in binders under their mattress.

**Comment by justinclift** [comment]
@revolter No worries, I'll get it done. :smile:

**Comment by justinclift** [comment]
Initial communication to upstream QScintilla in progress:

&nbsp; &nbsp; https://www.riverbankcomputing.com/pipermail/qscintilla/2016-October/001160.html

**Comment by ilovezfs** [comment]
@justinclift Thank you so much :bowing_man: You're great :)

**Comment by revolter** [comment]
It builds for me too (again ❤️).

**Comment by joshua-meng** [comment]
@ilovezfs @justinclift 

This patch builds ok but cause a serious bug of QScintilla.

Make ```QPointer<QsciLexer> lex;``` global instead of QsciScintilla's memeber would make lexer be singleton.

If you build an Application with multi-docs, different doc need to have different lexers (for different languages/apis/calltips), not share a global one. So lexer should always be a member of QsciScintilla's.

I made some changes based on this patch and it works fine without damage QScintilla's logic.

Here's the [LINK](https://github.com/joshua-meng/Lunar/commit/751e3f015ecbefe7fcb90f02092574711b6cec3d).

**Comment by justinclift** [comment]
Thanks @joshua-meng, that's an interesting thought.  Have you seen this?

* https://www.riverbankcomputing.com/pipermail/qscintilla/2016-October/001167.html
* https://github.com/opencor/opencor/commit/70f3944e36b8b95b3ad92106aeae2f511b3f0e90

Not sure which approach to go with. :smile:

**Comment by ilovezfs** [comment]
@agarny do you mind reviewing joshua-meng's patch (https://github.com/joshua-meng/Lunar/commit/751e3f015ecbefe7fcb90f02092574711b6cec3d) vs. yours?

@joshua-meng please test @agarny's patch and see if it works for you equally well.

>Not sure which approach to go with.

@justinclift if both patches work equally well
```
 24 additions and 0 deletions. 
```
vs.
```
78 additions and 67 deletions
```
...

**Comment by joshua-meng** [comment]
@ilovezfs 

Surely it works, I reviewed @agarny 's patch and found it totally same solution as mine. 

The only difference is I use std::unique_ptr to wrap the raw pointer of implementation struct.

Use ```#define lex (lexerStruct->object)``` is a good idea to avoid too much addtions/deletions.

**Comment by agarny** [comment]
@ilovezfs, @joshua-meng said it all.

**Comment by ilovezfs** [comment]
@joshua-meng @agarny so whose patch are we recommending to upstream, assuming upstream is even remotely paying attention? :see_no_evil:

**Comment by agarny** [comment]
@joshua-meng's patch requires C++11, which might be a problem for Phil Thompson (the person behind `QScintilla`), I don't know.

Anyway, Phil should already be aware of my patch since I posted a message on his mailing list (see https://www.riverbankcomputing.com/pipermail/qscintilla/2016-October/001167.html).

**Comment by justinclift** [comment]
@agarny Sounds like yours is the way to go then. :smile:

I'll look at reverting the initial patch we applied here then, and using yours instead. :smile:

**Comment by joshua-meng** [comment]
@agarny @justinclift @ilovezfs 
I found this bug basically caused by `extern "C++"`, when we remove `extern "C++"` from QsciScintilla, and move `QPointer<QsciLexer>` to cpp (because in QsciLexer.h, there is still `extern "C++"`, so we should move it out of .h file), it compiles ok.

Then I tried to remove all `#ifdef __APPLE__ extern "C++" { #endif` and `#ifdef __APPLE__ } #endif`  from QScintilla project (which appears in nearly all header files), compiles ok. links the output libraries to my App, works well too.

The conclusion is: This is all `extern "C++"` caused problem, it may be useful in early versions of XCode, but XCode 8 does not need that. I think QScintilla would not adopt PIMP patch to fix this bug because actually that was a magic skill of C++ to hide things to implementaion, temporary solution which just jump over the problem, not a perfect solution. I suppose QScintilla may remove `extern "C++"` later for the newest XCode.
